go/worker/registration/worker.go

```diff
@@ -109,19 +117,31 @@ func (r *Registration) doNodeRegistration() {
 			default:
 			}
 
-			return r.registerNode(epoch)
+			return w.registerNode(epoch)
 		}, off)
 	}
 
 	first := true
 	for {
 		select {
-		case <-r.stopCh:
+		case <-w.stopCh:
 			return
 		case epoch := <-ch:
+			// Check first if a clean halt was requested.
+			select {
+			case <-w.stopReqCh:
+				w.logger.Info("node registration halt requested")
+				w.Stop()
```

Comment:
Actually this needs to be changed to work as follows:
* When stop is requested the node should no longer update node registrations.
* But it must not shut down immediately after an epoch passes. Instead it must wait until the registry tells it that it has deregistered because a node can be registered until epoch X and if it is currently epoch X-10 then it needs to wait 10 epochs for deregistration before it is safe to shut down.